lib/acl.c

```diff
@@ -0,0 +1,243 @@
+	__attribute__((unused)) struct gatekeeper_if *iface)
+{
+	int i;
+	for (i = 0; i < num_pkts; i++)
+		rte_pktmbuf_free(pkts[i]);
+	return 0;
+}
+
+/*
+ * All IPv6 fields involved in classification; not all fields must
+ * be specified for every rule. Fields must be grouped into sets of
+ * four bytes, except for the first field.
+ */
+struct rte_acl_field_def ipv6_defs[NUM_FIELDS_IPV6] = {
+	{
+		.type = RTE_ACL_FIELD_TYPE_BITMASK,
+		.size = sizeof(uint8_t),
+		.field_index = PROTO_FIELD_IPV6,
+		.input_index = PROTO_FIELD_IPV6,
+		.offset = offsetof(struct ipv6_hdr, proto),
+	},
+	{
+		.type = RTE_ACL_FIELD_TYPE_MASK,
+		.size = sizeof(uint32_t),
+		.field_index = DST1_FIELD_IPV6,
+		.input_index = DST1_FIELD_IPV6,
+		.offset = offsetof(struct ipv6_hdr, dst_addr[0]),
+	},
+	{
+		.type = RTE_ACL_FIELD_TYPE_MASK,
+		.size = sizeof(uint32_t),
+		.field_index = DST2_FIELD_IPV6,
+		.input_index = DST2_FIELD_IPV6,
+		.offset = offsetof(struct ipv6_hdr, dst_addr[4]),
+	},
+	{
+		.type = RTE_ACL_FIELD_TYPE_MASK,
+		.size = sizeof(uint32_t),
+		.field_index = DST3_FIELD_IPV6,
+		.input_index = DST3_FIELD_IPV6,
+		.offset = offsetof(struct ipv6_hdr, dst_addr[8]),
+	},
+	{
+		.type = RTE_ACL_FIELD_TYPE_MASK,
+		.size = sizeof(uint32_t),
+		.field_index = DST4_FIELD_IPV6,
+		.input_index = DST4_FIELD_IPV6,
+		.offset = offsetof(struct ipv6_hdr, dst_addr[12]),
+	},
+	{
+		/* Enforce grouping into four bytes. */
+		.type = RTE_ACL_FIELD_TYPE_BITMASK,
+		.size = sizeof(uint32_t),
+		.field_index = TYPE_FIELD_ICMPV6,
+		.input_index = TYPE_FIELD_ICMPV6,
+		.offset = sizeof(struct ipv6_hdr) +
+			offsetof(struct icmpv6_hdr, type),
+	},
+};
+
+int
+register_ipv6_acl(struct ipv6_acl_rule *ipv6_rules, unsigned int num_rules,
+	acl_cb_func cb_f, struct gatekeeper_if *iface)
+{
+	unsigned int numa_nodes = get_net_conf()->numa_nodes;
+	unsigned int i;
+
+	if (iface->acl_func_count == GATEKEEPER_IPV6_ACL_MAX) {
+		RTE_LOG(ERR, GATEKEEPER, "acl: cannot install more ACL types on the %s iface\n",
+			iface->name);
+		return -1;
+	}
+
+	/* Assign a new ID for this rule type. */
+	for (i = 0; i < num_rules; i++)
+		ipv6_rules[i].data.userdata = iface->acl_func_count;
+
+	for (i = 0; i < numa_nodes; i++) {
+		int ret = rte_acl_add_rules(iface->ipv6_acls[i],
+			(struct rte_acl_rule *)ipv6_rules, num_rules);
+		if (ret < 0) {
+			RTE_LOG(ERR, ACL, "Failed to add ACL rules on the %s interface on socket %d\n",
+				iface->name, i);
+			return ret;
+		}
+	}
+
+	iface->acl_funcs[iface->acl_func_count] = cb_f;
+	iface->acl_func_count++;
+
+	return 0;
+}
+
+int
+process_ipv6_acl(struct gatekeeper_if *iface, unsigned int lcore_id,
+	struct acl_search *acl)
+{
+	struct rte_mbuf *pkts[iface->acl_func_count][GATEKEEPER_MAX_PKT_BURST];
+	int num_pkts[iface->acl_func_count];
+	unsigned int socket_id = rte_lcore_to_socket_id(lcore_id);
+	unsigned int i;
+	int ret;
+
+	if (!ipv6_if_configured(iface)) {
+		drop_ipv6_acl_pkts(acl->mbufs, acl->num, iface);
+		return 0;
```

Comment:
Whenever `process_ipv6_acl()` returns, it should zero `acl->num`. The reason for this is that `struct acl_search` may be defined outside of a loop. The only reason this problem has shown up is because `struct acl_search` is being defined inside loops.